src/InteractiveWindow/Editor/History.cs

```diff
@@ -155,82 +159,87 @@ private Entry Get(string pattern, int step)
             }
 
             return next;
-        }
+         }
 
-        internal Entry GetNext(string pattern = null)
+        internal Entry GetPrevious(string pattern)
         {
-            return Get(pattern, step: +1);
-        }
+            var startPos = _current;
+            Entry next;
+            next = MovePrevious(pattern);
+            if (next == null)
+            {
+                _current = startPos;
+                return null;
+            }
 
-        internal Entry GetPrevious(string pattern = null)
-        {
-            return Get(pattern, step: -1);
-        }
+            return next;
+         }
 
-        private Entry Move(string pattern, int step)
+        private Entry MoveNext(string pattern)
         {
-            Debug.Assert(step == -1 || step == +1);
+            if (Length == 0) return null;
+
+            bool wasUninitialized = (_current == -1);
+            if (wasUninitialized) return null;
 
-            bool wasLive = _live;
             _live = true;
 
-            if (Length == 0)
+            int end = Length - 1;
+            for (_current++; _current <= end; _current++)
             {
-                return null;
+                Entry entry;
+                if (TryMatch(pattern, out entry)) return entry;
             }
 
-            bool patternEmpty = string.IsNullOrWhiteSpace(pattern);
+            return null;
+        }
 
-            int start, end;
-            if (step > 0)
+        private Entry MovePrevious(string pattern)
+        {
+            if (Length == 0) return null;
+            bool wasLive = _live;
+            _live = true;
+
+            bool wasUninitialized = (_current == -1);
+            if (wasUninitialized)
             {
-                start = 0;
-                end = Length - 1;
+                _current = Length - 1;
+                Entry entry;
+                if (TryMatch(pattern, out entry)) return entry;
             }
-            else
+
+            int end = 0;
+            bool patternEmpty = string.IsNullOrWhiteSpace(pattern);
+            if (!wasLive && patternEmpty)
             {
-                start = Length - 1;
-                end = 0;
+                //return the current entry again ( handles case up, up, enter, up)
+                Entry entry;
+                if (TryMatch(pattern, out entry)) return entry;
             }
 
-            // no search in progress:
-            if (_current == -1)
+            for (_current--; _current >= end; _current--)
```

Comment:
Did you mean to leave the decl position empty?
Also, I'd be tempted to eliminate ```end``` and just use ```0```.